lego/connections.py

```diff
@@ -0,0 +1,134 @@
+"""
+Connections provides the connection to the components. A connection should provide simple functionality to
+the component, such as running a shell command or upload a file.
+Each connection should be based on different protocol, e.g. SSH or telnet.
+"""
+from __future__ import annotations
+from typing import Optional, Type
+from types import TracebackType
+import abc
+
+import plumbum
+import rpyc
+from rpyc.utils.zerodeploy import DeployedServer
+
+
+class BaseConnection(metaclass=abc.ABCMeta):
+    """
+    BaseConnection class is the base class of every connection to any type of components.
+    A connection should provide simple API to communicate and control the component.
+    """
+
+    def __enter__(self) -> BaseConnection:
+        """Allowing the use of 'with' statement with connections objects.
+
+        Returns:
+            Created class instance.
+        """
+
+        return self
+
+    def __exit__(
+            self,
+            exc_type: Optional[Type[BaseException]],
+            exc_value: Optional[BaseException],
+            traceback: Optional[TracebackType]) -> None:
+        """Closes the connection at the exit from the context manager.
+
+        Args:
+            exc_type: Exception type.
+            exc_value: Exception value.
+            traceback: Exception traceback.
+        """
+        self.close()
+
+    @abc.abstractmethod
+    def close(self) -> None:
+        """Closes all connections."""
+        raise NotImplementedError()
```

Comment:
Implementation of abstract methods is just:
```suggestion
        pass
```